src/urls/package-url-handler.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {lookupDependencyMapping} from '../manifest-converter';
+
+import {ConvertedDocumentFilePath, ConvertedDocumentUrl, OriginalDocumentUrl, PackageType} from './types';
+import {UrlHandlerInterface} from './url-handler-interface';
+import {getRelativeUrl} from './util';
+
+
+/**
+ * Handle URLs in a single "package-based" layout. This converter should be used
+ * to convert a single package, where all existing Bower dependencies are
+ * installed in a "bower_components/" sub-directory inside the main package
+ * directory.
+ */
+export class PackageUrlHandler implements UrlHandlerInterface {
+  readonly packageName: string;
+  readonly packageType: PackageType;
+
+  /**
+   * Helper function to check if a file URL is internal to the main package
+   * being converted (vs. a dependency).
+   */
+  static isUrlInternalToPackage(url: ConvertedDocumentUrl|OriginalDocumentUrl|
+                                ConvertedDocumentFilePath) {
+    return !url.startsWith('bower_components/') &&
+        !url.startsWith('./node_modules/');
+  }
+
+  constructor(packageName: string, packageType?: PackageType) {
+    this.packageName = packageName;
+    this.packageType = packageType || 'element';
+  }
+
+  /**
+   * Get the converted NPM name for a package, given the original document URL
+   * of any file that lives in that package.
+   */
+  getPackageNameForUrl(url: OriginalDocumentUrl) {
+    // If url contains no directories it must be internal. Explicitly check
+    // `basePackageName` here since it is needed below.
+    const urlParts = url.split('/');
+    const basePackageName = urlParts[1] as (string | undefined);
```

Comment:
Won't this fail for URLs with schemes? I think we should use the URL object, though it would require passing a base URL around: https://nodejs.org/api/url.html#url_constructor_new_url_input_base TODO/issue?

Reply:
No, this is only ever called on local documents / OriginalDocumentUrl is only  a local/relative URL.